src/sinks/s3_common/sink.rs

```diff
@@ -78,16 +231,77 @@ where
     }
 }
 
+struct PayloadBuilder {
+    framer: Framer,
+    compression: Compression,
+    encoder: crate::codecs::Encoder<Framer>,
+}
+
+struct Payload {
+    body: Bytes,
+    uncompressed_size: usize,
+}
+
+impl PayloadBuilder {
+    fn new(serializer: Serializer, framer: Framer, compression: Compression) -> Self {
+        // We don't actually use the serializer at all, we just need an `Encoder` for the batch
+        // prefix and suffix, which for some reason you can't get from the framer alone.
+        let encoder = crate::codecs::Encoder::<Framer>::new(framer.clone(), serializer);
```

Comment:
Does this need the explicit type bound since it is also specified in the struct?